#Imporint large model

In [ ]:
# Import spacy
import spacy

In [ ]:
# Directly download via spaCy (recommended)
!python -m spacy download en_core_web_lg

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 400.7/400.7 MB 3.4 MB/s eta 0:00:00
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')
⚠ Restart to reload dependencies
If you are in a Jupyter or Colab notebook, you may need to restart Python in
order to load all the package's dependencies. You can do this by selecting the
'Restart kernel' or 'Restart runtime' option.


In [ ]:
# Load the model
import spacy
nlp = spacy.load("en_core_web_lg")

In [ ]:
nlp('BRAC University').vector

array([ 0.312195  , -0.043325  ,  0.47918   , -0.295975  ,  0.288975  ,
       -0.132775  , -0.2498455 ,  0.13492998,  0.031175  ,  1.0480399 ,
       -0.620115  , -0.11784749,  0.15199   , -0.090975  ,  0.0691495 ,
        0.17890999, -0.135985  ,  0.50567   ,  0.30411   ,  0.168625  ,
        0.1912055 ,  0.3138885 , -0.164594  ,  0.0620325 ,  0.36967   ,
       -0.10243501, -0.616545  ,  0.276285  , -0.261145  ,  0.47197002,
       -0.218853  , -0.29312998,  0.36998   ,  0.2578705 , -0.0738975 ,
       -0.295485  ,  0.1879365 ,  0.35446   , -0.21843001,  0.0428545 ,
        0.14028999,  0.1042625 ,  0.16155   ,  0.156945  ,  0.247535  ,
       -0.113704  ,  0.1453335 , -0.42028502, -0.007158  , -0.27936   ,
       -0.4314825 ,  0.2003485 ,  0.288115  , -0.21426499, -0.01846001,
        0.13112   ,  0.2127105 , -0.54898   , -0.001457  ,  0.0560645 ,
       -0.2855    , -0.031115  ,  0.128135  , -0.13955599, -0.015605  ,
        0.09696101, -0.13087454,  0.50485003,  0.18434   ,  0.58

In [ ]:
# Doc
nlp('The quick brown fox jumps over the lazy dog.').vector.shape

(300,)

In [ ]:
# Single word
nlp('dog').vector.shape

(300,)

In [ ]:
# Identifying similar words
tokens = nlp('lion cat pet')

for token1 in tokens:
  for token2 in tokens:
    print(token1.text, token2.text, token1.similarity(token2))

lion lion 1.0
lion cat 0.5265437960624695
lion pet 0.39923766255378723
cat lion 0.5265437960624695
cat cat 1.0
cat pet 0.7505456805229187
pet lion 0.39923766255378723
pet cat 0.7505456805229187
pet pet 1.0


In [ ]:
tokens = nlp('love hate like')

for token1 in tokens:
  for token2 in tokens:
    print(token1.text, token2.text, token1.similarity(token2))

love love 1.0
love hate 0.6393099427223206
love like 0.6579040288925171
hate love 0.6393099427223206
hate hate 1.0
hate like 0.6574651598930359
like love 0.6579040288925171
like hate 0.6574651598930359
like like 1.0


In [ ]:
nlp.vocab.vectors.shape

(342918, 300)

In [ ]:
# Outside of vocab attribute checking
tokens = nlp('dog cat nargle')

for token1 in tokens:
  print(token1.text, token1.has_vector, token1.vector_norm, token1.is_oov)

dog True 7.0336733 False
cat True 6.6808186 False
nargle False 0.0 True


In [ ]:
# Outside of vocab attribute checking
tokens = nlp('dog cat karen')

for token1 in tokens:
  print(token1.text, token1.has_vector, token1.vector_norm, token1.is_oov)

dog True 7.0336733 False
cat True 6.6808186 False
karen True 5.5672445 False


In [ ]:
# Outside of vocab attribute checking
tokens = nlp('dog cat janet')

for token1 in tokens:
  print(token1.text, token1.has_vector, token1.vector_norm, token1.is_oov)

dog True 7.0336733 False
cat True 6.6808186 False
janet True 5.880165 False


In [ ]:
# NOTE: THIS IS A PERFECT CODE BUT IT'S NOT GIVING THE EXPECTED OUTPUT.
# MOST LIKELY, ISSUES RELATED WITH THE LARGE LANGUAGE MODEL

# King - Man + Woman = Queen.
# Importing spatial to perform arithmetic
from scipy import spatial

cosine_similarity = lambda vec1, vec2: 1 - spatial.distance.cosine(vec1, vec2)

# Grabbing their vector
king = nlp.vocab['king'].vector
man = nlp.vocab['man'].vector
woman  = nlp.vocab['woman'].vector

# King - Man + Woman = Queen --> new_vector similar queen, princess, highness

new_vector = king  - man + woman
computed_similarities = []

# Going all out in the entire vocabulary
# FOR ALL WORDSIN MY VOCAB

for word in nlp.vocab:
  if word.has_vector:
    if word.is_lower:
      if word.is_alpha: #numbers
        if word.is_stop == False:
          similarity = cosine_similarity(new_vector, word.vector)
          computed_similarities.append((word, similarity))

# Sorting in a decsending order
# Without the minus, it sorts in ascending order, showing the least similar words.
computed_similarities = sorted(computed_similarities, key = lambda item: -item[1])


# Printing the top 10 word from the tuple

# Grab the first word of the tuple, ask for it's text
# Then do it for tuple in computed_similarities
# Go through top 10, :10
print([t[0].text for t in computed_similarities[:10]])
# Expected output: ['queen', 'monarch', 'princess', 'royal', 'throne', ...]


['king', 'woman', 'lion', 'karen', 'janet', 'fox', 'brown', 'dare', 'cat', 'love']


#Sentiment Analysis (NLTK)

In [ ]:
import nltk

In [ ]:
nltk.download('vader_lexicon')

[nltk_data] Downloading package vader_lexicon to /root/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


True

In [ ]:
from nltk.sentiment.vader import SentimentIntensityAnalyzer

In [ ]:
sid = SentimentIntensityAnalyzer()

In [ ]:
review = "This is a good anime"

In [ ]:
sid.polarity_scores(review) # Only one +ve word

{'neg': 0.0, 'neu': 0.508, 'pos': 0.492, 'compound': 0.4404}

In [ ]:
review = "BEST ANIME! This was the best, most AWESOME anime MADE EVER IN HISTORY!!!"

In [ ]:
sid.polarity_scores(review) # It has many positive words, capitalized, even had '!!! marks'

{'neg': 0.0, 'neu': 0.393, 'pos': 0.607, 'compound': 0.9547}

In [ ]:
review = "This was the WORST anime!! YUCK! waste of my time!"

In [ ]:
sid.polarity_scores(review) # It has no positive word, rather bunch of negative words.

{'neg': 0.524, 'neu': 0.476, 'pos': 0.0, 'compound': -0.869}

##Amazon review

In [ ]:
import pandas as pd

In [ ]:
df = pd.read_csv('amazonreviews.tsv', sep='\t')

In [ ]:
df.head()

label                                             review
0   pos  Stuning even for the non-gamer: This sound tra...
1   pos  The best soundtrack ever to anything.: I'm rea...
2   pos  Amazing!: This soundtrack is my favorite music...
3   pos  Excellent Soundtrack: I truly like this soundt...
4   pos  Remember, Pull Your Jaw Off The Floor After He...

In [ ]:
df['label'].value_counts()

label
neg    5097
pos    4903
Name: count, dtype: int64

In [ ]:
# Data cleaning
df.dropna(inplace=True)

In [ ]:
blanks = []
for i, ib, rv in df.itertuples():
  # (index, label, review)
  if type(rv) == str:
    if rv.isspace():
      blanks.append(i)

In [ ]:
blanks #this list is empty so we don;t need to drop anything

[]

In [ ]:
# But if we had any, we would write

# df.drop(blanks, inplace=True)

In [ ]:
df.iloc[0]['review']

'Stuning even for the non-gamer: This sound track was beautiful! It paints the senery in your mind so well I would recomend it even to people who hate vid. game music! I have played the game Chrono Cross but out of all of the games I have ever played it has the best music! It backs away from crude keyboarding and takes a fresher step with grate guitars and soulful orchestras. It would impress anyone who cares to listen! ^_^'

In [ ]:
sid.polarity_scores(df.iloc[0]['review'])

{'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'compound': 0.9454}

In [ ]:
# It will take time
# It's running the whole polarity scores function on every single review
df['scored'] = df['review'].apply(lambda review: sid.polarity_scores(review))

In [ ]:
# Checking the head of the new dataframe
# We'll gt a new column scores containing the dictionary

df.head()

label                                             review  \
0   pos  Stuning even for the non-gamer: This sound tra...   
1   pos  The best soundtrack ever to anything.: I'm rea...   
2   pos  Amazing!: This soundtrack is my favorite music...   
3   pos  Excellent Soundtrack: I truly like this soundt...   
4   pos  Remember, Pull Your Jaw Off The Floor After He...   

                                              scored  
0  {'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'co...  
1  {'neg': 0.018, 'neu': 0.837, 'pos': 0.145, 'co...  
2  {'neg': 0.04, 'neu': 0.692, 'pos': 0.268, 'com...  
3  {'neg': 0.09, 'neu': 0.615, 'pos': 0.295, 'com...  
4  {'neg': 0.0, 'neu': 0.746, 'pos': 0.254, 'comp...

In [ ]:
# Creating a new compound column
# We'll set that equal to DF scores.
# Then we'll apply a lambda -> pass in the dict -> grab compound off of the dict
df['compound'] = df['scored'].apply(lambda d:d['compound'])

In [ ]:
# Ran into a small mistake
# Verifying DataFrame Columns

print(df.columns)  # Check if 'scores' exists


Index(['label', 'review', 'scored', 'compound'], dtype='object')


In [ ]:
df.head()

label                                             review  \
0   pos  Stuning even for the non-gamer: This sound tra...   
1   pos  The best soundtrack ever to anything.: I'm rea...   
2   pos  Amazing!: This soundtrack is my favorite music...   
3   pos  Excellent Soundtrack: I truly like this soundt...   
4   pos  Remember, Pull Your Jaw Off The Floor After He...   

                                              scored  compound  
0  {'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'co...    0.9454  
1  {'neg': 0.018, 'neu': 0.837, 'pos': 0.145, 'co...    0.8957  
2  {'neg': 0.04, 'neu': 0.692, 'pos': 0.268, 'com...    0.9858  
3  {'neg': 0.09, 'neu': 0.615, 'pos': 0.295, 'com...    0.9814  
4  {'neg': 0.0, 'neu': 0.746, 'pos': 0.254, 'comp...    0.9781

We got our compund scores here. Now let's add something; based off this compound score do a little bit of logic and say if it's greater than zero, then it's positive. If it's less than zero, it's negative. Then we'll compare these compound scores to the true labels we already know.

Lambda: take in a score, return back 'pos' if score >= 0, else return 'neg'

In [ ]:
df['compound_scores'] = df['compound'].apply(lambda score: 'pos' if score >= 0 else 'neg' )

In [ ]:
df.head()

label                                             review  \
0   pos  Stuning even for the non-gamer: This sound tra...   
1   pos  The best soundtrack ever to anything.: I'm rea...   
2   pos  Amazing!: This soundtrack is my favorite music...   
3   pos  Excellent Soundtrack: I truly like this soundt...   
4   pos  Remember, Pull Your Jaw Off The Floor After He...   

                                              scored  compound compound_scores  
0  {'neg': 0.088, 'neu': 0.669, 'pos': 0.243, 'co...    0.9454             pos  
1  {'neg': 0.018, 'neu': 0.837, 'pos': 0.145, 'co...    0.8957             pos  
2  {'neg': 0.04, 'neu': 0.692, 'pos': 0.268, 'com...    0.9858             pos  
3  {'neg': 0.09, 'neu': 0.615, 'pos': 0.295, 'com...    0.9814             pos  
4  {'neg': 0.0, 'neu': 0.746, 'pos': 0.254, 'comp...    0.9781             pos

In [ ]:
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

Comparing df of label (true label) with df of compound_scores (calculated by vader)

In [ ]:
accuracy_score(df['label'], df['compound_scores'])

0.7097

Let's print the classificatin report

In [ ]:
print(classification_report(df['label'], df['compound_scores']))

              precision    recall  f1-score   support

         neg       0.86      0.52      0.64      5097
         pos       0.64      0.91      0.75      4903

    accuracy                           0.71     10000
   macro avg       0.75      0.71      0.70     10000
weighted avg       0.75      0.71      0.70     10000



In [ ]:
print(confusion_matrix(df['label'], df['compound_scores']))

[[2629 2468]
 [ 435 4468]]


it's not performing well with the negative reviews. Also, the data set had some sarcastic reviews which is impossible for vader to detect propelry.

However, vader correctly identified positive/negative revioew roughly around 71%. It's not bad, like we really did a simple procedure here. But it's not good either comparing to some deep learning methods for sentiment analysis.